res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,715 @@
+    engine.setValue(group, "jog", newVal);
+};
+
+PioneerDDJ400.jogTouch = function(channel, _control, value) {
+    var deckNum = channel + 1;
+
+    // skip scratchmode if we adjust the loop points
+    if (this.loopin4beat[channel] || this.loopout[channel]) {
+        return;
+    }
+
+    // on touch jog with vinylmode enabled -> enable scratchmode
+    if (value !== 0 && this.vinylMode) {
+        engine.scratchEnable(deckNum, 720, 33+1/3, this.alpha, this.beta);
+    } else {
+        // on release jog (value === 0) disable pitch bend mode or scratch mode
+        engine.scratchDisable(deckNum);
+    }
+};
+
+///////////////////////////////////////////////////////////////
+//            HIGH RESOLUTION MIDI INPUT HANDLERS            //
+///////////////////////////////////////////////////////////////
+
+PioneerDDJ400.highResMSB = {
+    '[Channel1]': {},
+    '[Channel2]': {},
+    '[Channel3]': {},
+    '[Channel4]': {}
+};
+
+
+PioneerDDJ400.tempoSliderMSB = function(channel, control, value, status, group) {
+    PioneerDDJ400.highResMSB[group].tempoSlider = value;
+};
+
+PioneerDDJ400.tempoSliderLSB = function(channel, control, value, status, group) {
+    var fullValue = (PioneerDDJ400.highResMSB[group].tempoSlider << 7) + value;
+
+    engine.setValue(
+        group,
+        'rate',
+        ((0x4000 - fullValue) - 0x2000) / 0x2000
+    );
+};
+
+PioneerDDJ400.cycleTempoRange = function(_channel, _control, value, _status, group) {
+    if (value === 0) return; // ignore release
+
+    var currRange = engine.getValue(group, "rateRange");
+    var idx = 0;
+
+    for (var i = 0; i < this.tempoRanges.length; i++) {
+        if (currRange === this.tempoRanges[i]) {
+            idx = (i + 1) % this.tempoRanges.length;
+            break;
+        }
+    }
+    engine.setValue(group, "rateRange", this.tempoRanges[idx]);
+};
+
+
+var sortAsc = function(a, b) {
+    // returns 1 if a > b, -1 if a < b, and 0 otherwise
+    return (a > b) ? 1 : (b > a ? -1 : 0);
+};
+
+PioneerDDJ400.initCuePointsAndLoops = function(group) {
+    // create a list of positions in the track which can be selected
+    var points = [];
+
+    for (var padNum = 1; padNum <= 8; padNum++) {
+        points.push(engine.getValue(group, "hotcue_"+padNum+"_position"));
+    }
+    points.push(engine.getValue(group, "cue_point"));
+    points.push(engine.getValue(group, "loop_start_position"));
+    points.push(engine.getValue(group, "loop_end_position"));
+    points.sort(sortAsc); // sort asc
+    return points;
+};
+
+PioneerDDJ400.cueLoopCallLeft = function(_channel, _control, value, _status, group) {
+    if (value === 0) {
+        // ignore release
+        return;
+    }
+
+    var loopEnabled = engine.getValue(group, "loop_enabled");
+    if (loopEnabled) {
+        // loop halve
+        engine.setValue(group, "loop_scale", 0.5);
+    } else {
+        var currentPosition = engine.getValue(group, "playposition") - this.pointJumpSpace;
```

Comment:
ping https://github.com/mixxxdj/mixxx/pull/2312#discussion_r349979789

Reply:
@Be-ing This is a great example to ask for some general advice regarding advanced feature mappings.

The main question I have is *when to drop an original feature entirely from our mapping?*.

I think it's useful to specify rekordbucks's use case for the `<` and `>` arrow buttons to discuss this further:

1. During playback
  1.1 When in a loop, `<` halves the loop and `>` doubles it
  1.2 When not in a loop: **do nothing**
2. When playback is stopped
  2.1 In shift mode:
     *  `>` stores the loop or cue point (not hotcue) at the current position in memory
     * `<` deletes the current loop or cue from memory
   2.2 In normal mode
     * `>` jump forward to the next stored loop or cue
     * `<` jump back to the previous stored loop or cue

In our current mapping here, the above spec is a bit of a mess:
1.1. Is implemented 100%
1.2. Does the opposite of the original feature (i.e., jumps around during playback which is not good).
2.1. We can store a loop, but not delete it.
2.2. We can jump back and forth between the cue point, and up to 6 hot cues (or 5 plus one saved loop)

So clearly the more fancy features in 2. aren't met and are a bit haphazard. Dropping them makes total sense.

The reason I'm going into detail here is to find out whether you'd recommend to include the original feature if we implement it fully with a decent solution?

Reply:
> The main question I have is when to drop an original feature entirely from our mapping?.

When it's not very helpful, so complicated to use that it's impractical, or just doesn't make sense with Mixxx's design.

 Rekordbox's memory cues are a [weird legacy](https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/Rekordbox.20removable.20device.20library.20feature.20PR/near/186395716) from the technical limitations of embedded computing hardware from the mid 2000s. That's why no other DJ software implements them.

> Perhaps future work would include it completely?

I don't think there would be any point. Why bother scrolling through cue points when you already have instant access to all of them with dedicated buttons?

Reply:
Agreed and thanks for the guidelines.

In my personal mappings I use shift + `>` and `<` to jump forwards and backwards by a phrase. It's one of my most used mappings as I use it to prepare tracks with mixin cues as well as on the fly for the same purpose.

I suspect other people would find it useful, so happy to replace the memory cue stuff with it if you like?

Reply:
Shift + `<`/`>` for beatjump would be useful. Yeah, replace the memory cue stuff with that.